README.md
# Reversi_game


### Overview:
-------------
This project was done in a first year of my univeristy course. The goal was to enhance my experience with brown coding, debug this Reversi game and make it working as supposed and so I did! :) 

### Running the app:
--------------------
• To run the game just run the **main.java** class 



CHANGELOG
v1.0.10 [2020-05-04]:

    • Returned back {0,0} as a potential move in "DIRS" multidimensional array in isLegal method 


v1.0.9 [2020-05-03]:

    • Added missing part of the string in the test of isLegal method
    • Restored back condition in "else if" statement in getFinalScore method
    • Fixed "while" loop in flipPieces method by changing operator && to ||


v1.0.8 [2020-04-25]:

    • Fixed the test for findPotentialMoves method by replacing size of gray cells from 5 to 4
    • Deleted {0,0} as a potential move from "DIRS" multidimensional array in isLegal method 
    • Added missing condition (d_row < BOARD_SIZE) to the "if" statement in "for each" loop in isLegal method
    • Restored isLegal method to one of the previous implentations (replaced "do-while" loop to "while" loop with 
    its previous conditions, added "if" statement which I deleted before)


v1.0.7 [2020-04-22]:

    • Replaced "CellStatus.Dark" to "cellstatus" parameter in an ArrayList "potentialMoves" in generateOpponent method 
    • Added "this.cells[d_row][d_col].setValue(colour)" statement to the body of while loop in flipPieces method


v1.0.6 [2020-04-08]:

    • Changed operators in nested "for" loops in findPotentialMoves method to <
    • Replaced "while" loop to to "do-while loop" in isLegal method
    • Deleted {0,0} as a potential move from DIRS multidimensional array in isLegal method 
    • Fixed getRow method which was returning column instead of row
    • Fixed problem with ArrayIndexOutOfBoundsException error by changing "d_col <= BOARD_SIZE" to "d_col < BOARD_SIZE" 
    and "d_row <= BOARD_SIZE" to "d_row < BOARD_SIZE" in "if" statements in isLegal method
    • Changed conditions in "if" statement in "for each loop" in isLegal method
    • Delete one "if" statement from "for each loop" in isLegal method
   

v1.0.5 [2020-04-03]:

    • Changed wrong indictaion of column in "dir" array (was dir[0] instead of dir[1] which represents columns) 
    in isLegal method
    • Added missing condition "d_row <= BOARD_SIZE" to the "if" statement in isLegal method


v1.0.4 [2020-03-25]:

    • Deleted redundant "break" statement in default case in setValue method
    • Initialised "cells" attribute in a constructor in MoveChecker class
    • Changed operators in "if" statements of nested for loops in findPotentialMoves method from <= to ==
    • Returned back condition "CellStatus.EMPTY" in "if" statement in findPotentialMoves method
    • Changed operator in "if" statement in isLegal method
    
    
v1.0.3 [2020-03-23]:

    • Modified "if" statement in findPotentialMoves method
    • Changed operators in "if" statement inside of the for loop in isLegal method
    
    
v1.0.2 [2020-03-19]:

    • Changed operators in "if" statements of nested for loops in findPotentialMoves method from <= to <
    

v1.0.1 [2020-03-17]:

    • Erased condition to meet in "else" statement in getFinalScore method 
    • Added missing semicolon
    • Defined "cells" parameter in javadoc in isLegal method
    • Fixed "if" statement in getFinalScore method by modifying "else" condition
    
    
v1.0 [2020-03-16]:

    • First major release  



Reversi/src/uk/ac/ncl/tests/MoveCheckerTest.java
package uk.ac.ncl.tests;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import uk.ac.ncl.entity.Cell;
import uk.ac.ncl.entity.CellStatus;
import uk.ac.ncl.game.MoveChecker;

import javax.swing.*;

import java.util.ArrayList;

import static org.junit.jupiter.api.Assertions.*;
import static uk.ac.ncl.Constants.BOARD_SIZE;

class MoveCheckerTest {

    private MoveChecker moveChecker;
    private Cell[][] cells;

    @BeforeEach
    void setUp() {
        cells = new Cell[BOARD_SIZE][BOARD_SIZE];
        moveChecker = new MoveChecker(cells);
        for (int i = 0; i < BOARD_SIZE; i++) {
            for (int j = 0; j < BOARD_SIZE; j++) {
                CellStatus cellStatus = CellStatus.EMPTY;
                Cell tempEl = new Cell(cellStatus, new JButton(), i, j);
                if ((i == 3 && j == 4) || (i == 4 && j == 3)) {
                    tempEl.setValue( CellStatus.LIGHT);
                } else if ((i == 4 && j == 4) || (i == 3 && j == 3)) {
                    tempEl.setValue( CellStatus.DARK);
                }
                cells[i][j] = tempEl;
            }
        }
    }

    @Test
    void generateOpponentDark() {
        Cell cell = moveChecker.generateOpponent(CellStatus.DARK);

        assertTrue(cells[2][4] == cell || cells[3][5] == cell ||
                cells[4][2] == cell || cells[5][3] == cell);
        assertTrue(cell.getMove() != null);
        assertTrue(cell.getMove().getMoves().size() == 1);
        assertTrue(cell.getMove().getScore() == 1);
    }

    @Test
    void generateOpponentLight() {
        Cell cell = moveChecker.generateOpponent(CellStatus.LIGHT);

        assertTrue(cells[2][3] == cell || cells[3][2] == cell ||
                cells[4][5] == cell || cells[5][4] == cell);
        assertTrue(cell.getMove() != null);
        assertTrue(cell.getMove().getMoves().size() == 1);
        assertTrue(cell.getMove().getScore() == 1);
    }

    @Test
    void flipCheckers() {
        cells[5][3].isLegal(CellStatus.DARK, this.cells);
        moveChecker.flipPieces(cells[5][3], CellStatus.DARK);
        assertTrue(cells[4][3].getValue() == CellStatus.DARK); // nie wiem co z tym zrobic
        assertTrue(cells[4][3].getMove() == null);
    }

    @Test
    void findPotentialMoves() {
        ArrayList<Cell> grayCells = moveChecker.findPotentialMoves(CellStatus.DARK);
        assertTrue(grayCells.size() == 4);
        assertTrue(grayCells.contains(cells[2][4]));
        assertTrue(grayCells.contains(cells[3][5]));
        assertTrue(grayCells.contains(cells[4][2]));
        assertTrue(grayCells.contains(cells[5][3]));
    }

    @Test
    void colourPieces() {
        CellStatus testColour = CellStatus.GRAY;
        ArrayList<Cell> piecesToColour = new ArrayList<Cell>();
        piecesToColour.add(cells[2][4]);
        piecesToColour.add(cells[3][5]);
        piecesToColour.add(cells[4][2]);
        piecesToColour.add(cells[5][3]);
        moveChecker.colourPieces(piecesToColour, testColour);
        assertTrue(cells[2][4].getValue() == testColour);
        assertTrue(cells[3][5].getValue() == testColour);
        assertTrue(cells[4][2].getValue() == testColour);
        assertTrue(cells[5][3].getValue() == testColour);
    }

    @Test
    void getFinalScore() {
        String result = "The game is over. It is a draw. Each player has " + 2 + " pieces";
        assertTrue(result.equals(moveChecker.getFinalScore()));
    }

    @Test
    void removeMoves() {
        ArrayList<Cell> grayCells = moveChecker.findPotentialMoves(CellStatus.DARK);
        assertTrue(cells[2][4].getMove() != null);
        assertTrue(cells[3][5].getMove() != null);
        assertTrue(cells[4][2].getMove() != null);
        assertTrue(cells[5][3].getMove() != null);
        moveChecker.removeMoves(grayCells);
        assertTrue(cells[2][4].getMove() == null);
        assertTrue(cells[3][5].getMove() == null);
        assertTrue(cells[4][2].getMove() == null);
        assertTrue(cells[5][3].getMove() == null);
    }

    @Test
    void testIsLegal(){
        assertTrue(cells[5][3].isLegal(CellStatus.DARK, this.cells));
        assertTrue(cells[5][3].getMove().getMoves().size() == 1);
        assertTrue(cells[5][3].getMove().getMoves().get(0).getCell().getColumn() == 3);
        assertTrue(cells[5][3].getMove().getMoves().get(0).getCell().getRow() == 3);
        assertTrue(cells[5][3].getMove().getScore() == 1);
    }
}

Reversi/src/uk/ac/ncl/entity/Cell.java
/**
 * @author Kostiantyn Potomkin
 * @version 0.9
 * @since 05-03-2020
 */
package uk.ac.ncl.entity;

import javax.swing.*;
import javax.swing.border.LineBorder;
import java.awt.*;
import java.util.ArrayList;

import static uk.ac.ncl.Constants.BOARD_SIZE;

/**
 * Represents each cell of the playing board.
 */
public class Cell {

    /**
     * Current cell status.
     */
    private CellStatus value;
    /**
     * Links cell to the corresponding UI element.
     */
    private JButton jButton;
    /**
     * Cell's row
     */
    private int row;
    /**
     *  Cell's column
     */
    private int column;

    /**
     *  Potential moves of the piece
     */
    private Move move;

    public Cell(CellStatus value, JButton jButton, int row, int column){
        this.value = value;
        this.jButton = jButton;
        this.row = row;
        this.column = column;
        this.move = null;
    }

    public CellStatus getValue() {
        return value;
    }

    /**
     *   Changes button's design to have an effect of the "pressed" button
     */
    public void colourTemp(Color colour, boolean isPressed){
        this.jButton.setBackground(colour);
        if (isPressed) {
            this.jButton.setBorderPainted(true);
            this.jButton.setBorder(new LineBorder(Color.RED));
        }
        else {
            this.jButton.setBorderPainted(false);
        }
    }

    /**
     *   Updates the status of the cell
     */
    public void setValue(CellStatus value) {
        this.value = value;
        switch (value){
            case EMPTY:
                this.jButton.setBackground(new Color(820000));
                break;
            case LIGHT:
                this.jButton.setBackground(Color.WHITE);
                break;
            case DARK:
                this.jButton.setBackground(Color.BLACK);
                break;
            case GRAY:
                this.jButton.setBackground(Color.GRAY);
                break;
            default:
        }
    }

    public int getRow() {
        return row;
    }

    public int getColumn() {
        return column;
    }

    public Move getMove() {
        return move;
    }

    public void setMove(Move move) {
        this.move = move;
    }

    /**
     * Checks whether there exists a legal move for the piece.
     * If such a move exists, returns true and adds information to the piece.
     * @param colour - colour of the current player
     * @param cells - current cell on the board
     * @return whether move is possible for the piece. If this is the case, then possible moves are stored in Piece.
     */
    public boolean isLegal(CellStatus colour, Cell[][] cells){
        CellStatus opponent = colour == CellStatus.LIGHT ? CellStatus.DARK : CellStatus.LIGHT;
        boolean isLegal = false;
        int score = 0;
        ArrayList<DirectedMove> moves = new ArrayList<DirectedMove>();
        int[][] DIRS = {{-1,-1}, {-1,0}, {0,0}, {-1,1}, {0,1}, {0,-1}, {1,1}, {1,0}, {1,-1}};


        for (int[] dir : DIRS){
            int temp_score = 0;
            int d_row = this.getRow() + dir[0];
            int d_col = this.getColumn() + dir[1];
            if (0 <= d_col &&  d_col < BOARD_SIZE && 0 <=  d_row && d_row < BOARD_SIZE
                    && cells[d_row][d_col].getValue() != CellStatus.EMPTY
                    && cells[d_row][d_col].getValue() == opponent) {
                while (true){
                    d_row += dir[0];
                    d_col += dir[1];
                    temp_score += 1;

                    if (0 <= d_col &&  d_col < BOARD_SIZE && 0 <= d_row &&  d_row < BOARD_SIZE
                            && cells[d_row][d_col].getValue() != CellStatus.EMPTY){
                        if (cells[d_row][d_col].getValue() == colour){
                            isLegal = true;
                            score += temp_score;
                            moves.add(new DirectedMove(cells[d_row][d_col], dir));
                        }
                    }
                    else {
                        break;
                    }

                }

            }
        }
        Move move = new Move(moves, score);
        this.setMove(move);
        return isLegal;
    }
}


Reversi/src/uk/ac/ncl/game/MoveChecker.java
/**
 * @author Kostiantyn Potomkin
 * @version 0.9
 * @since 05-03-2020
 */
package uk.ac.ncl.game;

import uk.ac.ncl.entity.CellStatus;
import uk.ac.ncl.entity.DirectedMove;
import uk.ac.ncl.entity.Cell;
import java.util.ArrayList;



import static uk.ac.ncl.Constants.*;

/**
 *
 * Main logic of the game.
 *
 */
public class MoveChecker {

    private Cell[][] cells;

    public MoveChecker(Cell[][] cells) {
        this.cells = cells;
    }

    /**
     * Generates a move of the opponent
     *
     * @param cellStatus - colour of the opponent (light)
     * @return a piece to make a move
     */
    public Cell generateOpponent(CellStatus cellStatus) {
        ArrayList<Cell> potentialMoves = findPotentialMoves(cellStatus);
        int max_score = 0;
        Cell opponentsMove = null;
        for (int i = 0; i < potentialMoves.size(); i++){
            if (potentialMoves.get(i).getMove().getScore() > max_score){
                opponentsMove = potentialMoves.get(i);

            }
        }
        return opponentsMove;
    }

    /**
     * Flips pieces between selected piece in directions of valid moves.
     *
     * @param cell - piece we have just put on the board
     * @param colour - colour of the current player
     */
    public void flipPieces(Cell cell, CellStatus colour) {
        cell.colourTemp(colour == OPPONENTS_CELL_STATUS ? OPPONENTS_COLOUR : PLAYERS_COLOUR, false); // imitates the red border to the piece which I put
        for (DirectedMove move : cell.getMove().getMoves()) {
            int[] dir = move.getDirection();
            int d_row = cell.getRow();
            int d_col = cell.getColumn();

            while (d_col != move.getCell().getColumn() || d_row != move.getCell().getRow()) {
                this.cells[d_row][d_col].setValue(colour);
                    d_row += dir[0];
                    d_col += dir[1];

            }

        }
    }

    /**
     * Returns potential moves on the board for the specified colour
     *
     * @param colour - colour of the current player
     * @return pieces for which there exist valid moves
     */
    public ArrayList<Cell> findPotentialMoves(CellStatus colour) {
        ArrayList<Cell> potentialMoves = new ArrayList<Cell>();
        for (int i = 0; i < BOARD_SIZE; i++) {
            for (int j = 0; j < BOARD_SIZE; j++) {
                if (this.cells[i][j].getValue() == CellStatus.EMPTY){
                    if (this.cells[i][j].isLegal(colour, cells)){
                        potentialMoves.add(this.cells[i][j]);
                    }
                }
            }
        }
        return potentialMoves;

    }


    /**
     * Updates information about pieces
     *
     * @param cells - pieces to update
     * @param colour - new colour
     */
    public void colourPieces(ArrayList<Cell> cells, CellStatus colour) {
        for (Cell cell : cells){
            cell.setValue(colour);
        }
    }

    /**
     * Sums up results of the game
     *
     * @return string with the results of the game
     */
    public String getFinalScore() {
        int lights = 0;
        int darks = 0;
        for (int row = 0; row < BOARD_SIZE; row++) {
            for (int column = 0; column < BOARD_SIZE; column++) {
                if (this.cells[row][column].getValue() == CellStatus.LIGHT) {
                    lights++;
                } else if (this.cells[row][column].getValue() == CellStatus.DARK) {
                    darks++;
                }

            }
        }
        if (darks == lights)
            return "The game is over. It is a draw. Each player has " + darks + " pieces";

        String winner = lights > darks ? "Light" : "Dark";
            return "The game is over. "
                    + winner + " has won with the result: Dark - " + darks + " Light - " + lights;
    }

    /**
     * Cleans up potential moves from the board cells
     *
     * @param grayCells - pieces with valid moves
     */
    public void removeMoves(ArrayList<Cell> grayCells){
        if (grayCells != null){
            for (int i = 0; i < grayCells.size(); i++){
                grayCells.get(i).setMove(null);
            }
        }
    }


}
